### PR TITLE
Allow retries in downloading artifacts

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Part of the fix for hazelcast-enterprise#5023

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/5569

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java [hunk 1]
```diff
@@ -61,94 +55,61 @@ public enum Artifact {
...
-            fos = new FileOutputStream(targetFile);
-            drainTo(is, fos);
-            targetFile.deleteOnExit();
-            return targetFile;
-        } catch (IOException e) {
-            throw rethrowGuardianException(e);
-        } finally {
-            closeResource(fos);
-            closeResource(is);
+            int code = builder.start().waitFor();
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:90]:
```suggestion
            int code = builder.start().waitFor(120, TimeUnit.SECONDS);
```

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java [hunk 2]
```diff
@@ -61,94 +55,61 @@ public enum Artifact {
...
+            checkState(code == 0, "Maven dependency:get failed");
+        } catch (InterruptedException | IOException e) {
+            throw new RuntimeException("Problem in invoking Maven dependency:get", e);
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:93]:
```suggestion
            throw new RuntimeException("Problem in invoking Maven dependency:get " + artifact + ":" + version , e);
```

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java [hunk 3]
```diff
@@ -61,94 +55,61 @@ public enum Artifact {
...
 
     private static void logWarningForArtifactDownload(Artifact artifact, String version) {
         if (isOnJenkins()) {
             return;
         }
+        LOGGER.warning("Hazelcast binaries for version " + version
+                + (artifact.enterprise ? " EE " : " ")
+                + "will be downloaded from a remote repository. You can speed up the compatibility tests by "
+                + "installing the missing artifacts in your local maven repository so they don't have to be "
+                + "downloaded each time:\n $ " + buildMavenCommand(artifact, version));
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:105]: 👍

File: pom.xml [hunk 4]
```diff
@@ -573,6 +573,9 @@
                             <artifact>com.hazelcast:hazelcast:4.0.3</artifact>
                             <artifact>com.hazelcast:hazelcast:4.0.3:jar:tests</artifact>
 
+                            <artifact>com.hazelcast:hazelcast:5.2.0</artifact>
+                            <artifact>com.hazelcast:hazelcast:5.2.0:jar:tests</artifact>
```
Comment (ldziedziul) [inline pom.xml:577]:
```suggestion
                            <artifact>com.hazelcast:hazelcast:${hazelcast.previous.version}</artifact>
                            <artifact>com.hazelcast:hazelcast:${hazelcast.previous.version}:jar:tests</artifact>
```

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java [hunk 5]
```diff
@@ -87,10 +88,10 @@ private static void downloadArtifact(Artifact artifact, String version) {
         ProcessBuilder builder = new ProcessBuilder(buildMavenCommand(artifact, version).split(" "))
                 .inheritIO();
         try {
-            int code = builder.start().waitFor();
-            checkState(code == 0, "Maven dependency:get failed");
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java]: What do you think about checking `process.exitValue()==0` as well?

Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java]: Good idea, added

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java [hunk 6]
```diff
@@ -61,94 +56,63 @@ public enum Artifact {
...
-            targetFile.deleteOnExit();
-            return targetFile;
-        } catch (IOException e) {
-            throw rethrowGuardianException(e);
-        } finally {
-            closeResource(fos);
-            closeResource(is);
+            Process process = builder.start();
+            boolean successful = process.waitFor(120, TimeUnit.SECONDS);
+            checkState(successful, "Maven dependency:get failed");
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:93]:
```suggestion
            checkState(successful, "Maven dependency:get timed out");
```

**Review by ldziedziul — APPROVED** [review]

**Pull request merged**